internal/metrics/reporter.go

```diff
@@ -0,0 +1,326 @@
+		return nil, err
+	}
+	return &metricspb.IngestRequest{
+		IdempotencyKey: idempotencyKey,
+		Reporter:       r.collectorReporter,
+	}, nil
+}
+
+func (r *Reporter) addFloat(key string, value float64, kind metricspb.MetricKind, intervals int64) *metricspb.MetricPoint {
+	return &metricspb.MetricPoint{
+		Kind:       kind,
+		MetricName: key,
+		Labels:     r.labels,
+		Value: &metricspb.MetricPoint_DoubleValue{
+			DoubleValue: value,
+		},
+		Start: &types.Timestamp{
+			Seconds: r.Start.Unix(),
+			Nanos:   int32(r.Start.Nanosecond()),
+		},
+		Duration: &types.Duration{
+			Seconds: int64(DefaultReporterMeasurementDuration.Seconds()) * intervals,
+		},
+	}
+}
+
+// Measure takes a snapshot of system metrics and sends them
+// to a LightStep endpoint.
+func (r *Reporter) Measure(ctx context.Context, intervals int64) error {
+	start := time.Now()
+	r.Start = start
+	ctx, cancel := context.WithTimeout(ctx, r.timeout)
+	defer cancel()
+
+	m, err := Measure(ctx, 0*time.Second)
+	if err != nil {
+		return err
+	}
+
+	pb, err := r.prepareRequest(m)
+	if err != nil {
+		return err
+	}
+
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.cpu.user", m.ProcessCPU.User-r.stored.ProcessCPU.User, metricspb.MetricKind_COUNTER, intervals))
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.cpu.sys", m.ProcessCPU.System-r.stored.ProcessCPU.System, metricspb.MetricKind_COUNTER, intervals))
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.gc.count", float64(m.GarbageCollector.NumGC-r.stored.GarbageCollector.NumGC), metricspb.MetricKind_COUNTER, intervals))
+
+	pb.Points = append(pb.Points, r.addFloat("mem.available", float64(m.Memory.Available), metricspb.MetricKind_GAUGE, intervals))
+	pb.Points = append(pb.Points, r.addFloat("mem.total", float64(m.Memory.Used), metricspb.MetricKind_GAUGE, intervals))
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.mem.heap_alloc", float64(m.Memory.HeapAlloc), metricspb.MetricKind_GAUGE, intervals))
+
+	for label, cpu := range m.CPU {
+		pb.Points = append(pb.Points, r.addFloat("cpu.sys", cpu.System-r.stored.CPU[label].System, metricspb.MetricKind_COUNTER, intervals))
+		pb.Points = append(pb.Points, r.addFloat("cpu.user", cpu.User-r.stored.CPU[label].User, metricspb.MetricKind_COUNTER, intervals))
+		pb.Points = append(pb.Points, r.addFloat("cpu.total", cpu.Total-r.stored.CPU[label].Total, metricspb.MetricKind_COUNTER, intervals))
+		pb.Points = append(pb.Points, r.addFloat("cpu.usage", cpu.Usage-r.stored.CPU[label].Usage, metricspb.MetricKind_COUNTER, intervals))
+	}
+	for label, nic := range m.NIC {
+		pb.Points = append(pb.Points, r.addFloat("net.bytes_recv", float64(nic.BytesReceived-r.stored.NIC[label].BytesReceived), metricspb.MetricKind_COUNTER, intervals))
+		pb.Points = append(pb.Points, r.addFloat("net.bytes_sent", float64(nic.BytesSent-r.stored.NIC[label].BytesSent), metricspb.MetricKind_COUNTER, intervals))
+	}
+
+	err = r.send(ctx, pb)
+	if err != nil {
+		return err
+	}
+
+	r.stored = m
```

Comment:
Suggestion: access these mutable fields behind a `sync.RWMutex`. My brain's pretty fuzzy right now, though; I'm trying to figure out what will happen if, e.g., a previous request is still getting retried when a new one comes in (especially if the timeout and/or metric duration timeouts have been overridden).

Reply:
From what I was seeing, the context will timeout and return an error before another request is created